crates/wast/src/wast.rs

```diff
@@ -542,15 +398,78 @@ fn is_arithmetic_f64_nan(bits: u64) -> bool {
     (bits & AF64_NAN) == AF64_NAN
 }
 
-fn values_equal(v1: &Val, v2: &Val) -> Result<bool> {
-    Ok(match (v1, v2) {
-        (Val::I32(a), Val::I32(b)) => a == b,
-        (Val::I64(a), Val::I64(b)) => a == b,
-        // Note that these float comparisons are comparing bits, not float
-        // values, so we're testing for bit-for-bit equivalence
-        (Val::F32(a), Val::F32(b)) => a == b,
-        (Val::F64(a), Val::F64(b)) => a == b,
-        (Val::V128(a), Val::V128(b)) => a == b,
-        _ => bail!("don't know how to compare {:?} and {:?} yet", v1, v2),
+fn matches(actual: &Val, expected: &AssertExpression) -> Result<bool> {
+    Ok(match (actual, expected) {
+        (Val::I32(a), AssertExpression::I32(b)) => a == b,
+        (Val::I64(a), AssertExpression::I64(b)) => a == b,
+        (Val::F32(a), AssertExpression::F32(b)) => f32_matches(*a, b),
+        (Val::F64(a), AssertExpression::F64(b)) => f64_matches(*a, b),
+        (Val::V128(a), AssertExpression::V128(b)) => v128_matches(*a, b),
+        _ => bail!(
+            "don't know how to compare {:?} and {:?} yet",
+            actual,
+            expected
+        ),
     })
 }
+
+fn f32_matches(actual: u32, expected: &NanPattern<Float32>) -> bool {
+    match expected {
+        NanPattern::CanonicalNan => is_canonical_f32_nan(actual),
+        NanPattern::ArithmeticNan => is_arithmetic_f32_nan(actual),
+        NanPattern::Value(expected_value) => actual == expected_value.bits,
+    }
+}
+
+fn f64_matches(actual: u64, expected: &NanPattern<Float64>) -> bool {
+    match expected {
+        NanPattern::CanonicalNan => is_canonical_f64_nan(actual),
+        NanPattern::ArithmeticNan => is_arithmetic_f64_nan(actual),
+        NanPattern::Value(expected_value) => actual == expected_value.bits,
+    }
+}
+
+fn to_le_bytes(v: &V128Pattern) -> Vec<u8> {
+    match v {
+        V128Pattern::I8x16(w) => w.iter().map(|&i| i as u8).collect(),
+        V128Pattern::I16x8(w) => w.iter().flat_map(|i| i.to_le_bytes().to_vec()).collect(),
+        V128Pattern::I32x4(w) => w.iter().flat_map(|i| i.to_le_bytes().to_vec()).collect(),
+        V128Pattern::I64x2(w) => w.iter().flat_map(|i| i.to_le_bytes().to_vec()).collect(),
+        _ => panic!("Unable to convert a float pattern to bytes (may contain NaN patterns)"),
```

Comment:
Should this be a `bail!` with corresponding `Result`s in a few places?

Reply:
Yeah where possible we want to use `Result` to improve the error messages coming out of failed or unsupported tests